torch/_functorch/aot_autograd.py

```diff
@@ -1595,6 +1587,37 @@ def format_guard_bug_msg(aot_config, expected):
     )
 
 
+def remove_dupe_metadata(
+    m: ViewAndMutationMeta,
+    keep_arg_mask: List[bool],
+) -> ViewAndMutationMeta:
+    assert len(m.input_info) == len(keep_arg_mask)
+    # Easy invariant: the first argument should never be a dupe (it will be kept)
+    assert len(keep_arg_mask) > 0 and keep_arg_mask[0]
+    dupe_to_dedup_idx = [0]
+    for i, b in enumerate(keep_arg_mask[1:]):
+        if b:
+            dupe_to_dedup_idx.append(dupe_to_dedup_idx[-1] + 1)
+        else:
+            dupe_to_dedup_idx.append(dupe_to_dedup_idx[-1])
+
+    return ViewAndMutationMeta(
+        input_info=[x for i, x in enumerate(m.input_info) if keep_arg_mask[i]],
+        # requires_grad_info consists of (mutated_inputs, forward_outputs).
+        # Need to remove only the duplicate entries that correspond to the mutated inputs.
+        requires_grad_info=[
+            x for i, x in enumerate(m.requires_grad_info)
+            if i >= len(m.mutated_inp_indices) or keep_arg_mask[m.mutated_inp_indices[i]]],
+        # For outputs that are views of inputs, we store the index of the input that the output
+        # was generated from. Need to update that index to account for removed dupes.
+        output_info=[
+            OutputAliasInfo(o.output_type, None if o.base_idx is None else dupe_to_dedup_idx[o.base_idx])
+            for o in m.output_info],
+        num_intermediate_bases=m.num_intermediate_bases,
+        keep_input_mutations=m.keep_input_mutations,
+        traced_tangents=m.traced_tangents,
```

Comment:
As a debug option, it would be nice to be able rerun the analysis from scratch and compare that the manually deduped metadata agrees with the manual analysis.

Reply:
yep definitely a good idea

Reply:
debug asserts are currently on by default - it might make sense to turn them off by default now (and only turn them on by default for the aot_eager backend?). In my next PR, I do the same thing with synthetic bases - so debug asserts would mean we trace the forward with functionalization 2 extra times.

Reply:
(for now I'm making that change, to only turn them on in our test_aotdispatch tests + when using aot_eager)